File: apple/watch/OlympiadsRootView.swift
import SwiftUI
import ScienceCore

/// Top-level screen. Owns the activity list, the current subject
/// filter, and the loading/error state.
///
/// Differs from the iPhone `OlympiadsView` in two ways:
/// - The subject filter is persisted via `@AppStorage` so it sticks
///   across launches. The watch gets opened from the glance stack
///   many times a day — randomizing every time (as the iPhone does)
///   would be actively annoying.
/// - On cold start, the view reads the last successful response
///   from `ActivityCache` before the network call returns. The watch
///   radio is slow and flaky; a local cache makes the first frame
///   useful instead of a spinner. The cached list is then overwritten
///   by whatever comes back from the API, but only on success.
struct OlympiadsRootView: View {
    @State private var activities: [Activity] = []
    /// Empty string is the "All" sentinel — `@AppStorage` has no
    /// native `Optional<String>` support, so we bridge to nil via
    /// `effectiveFilter`.
    @AppStorage("watchOlympiadsFilter") private var storedFilter: String = ""
    @State private var initialLoading = true
    @State private var errorMessage: String?
    @State private var showingFilter = false

    private var effectiveFilter: String? {
        storedFilter.isEmpty ? nil : storedFilter
    }

    /// Two-way binding used by the filter sheet. Reads from
    /// `storedFilter`; writes round-trip through the same @AppStorage
    /// key so the choice persists immediately.
    private var filterBinding: Binding<String?> {
        Binding(
            get: { effectiveFilter },
            set: { storedFilter = $0 ?? "" }
        )
    }

    var body: some View {
        NavigationStack {
            Group {
                if initialLoading {
                    ProgressView()
                } else if activities.isEmpty, let errorMessage {
                    WatchErrorView(message: errorMessage)
                } else {
                    OlympiadsListView(
                        yearGroups: yearGroups,
                        activeFilter: effectiveFilter
                    )
                }
            }
            .navigationTitle("Olympiads")
            .toolbar {
                ToolbarItem(placement: .topBarTrailing) {
                    Button {
                        showingFilter = true
                    } label: {
                        Image(systemName: "line.3.horizontal.decrease.circle")
                    }
                    .accessibilityLabel("Filter by subject")
                }
            }
            .sheet(isPresented: $showingFilter) {
                SubjectFilterSheet(selection: filterBinding)
            }
            .task { await load() }
            .refreshable { await refresh() }
        }
    }

    // MARK: - Grouping

    private var yearGroups: [ActivityGrouping.YearGroup] {
        let filtered = ActivityGrouping.filtered(activities, subject: effectiveFilter)
        return ActivityGrouping.groupedByYear(filtered)
    }

    // MARK: - Loading

    /// Cold-start load path. First, synchronously hydrate from the
    /// on-disk cache so the user sees content immediately. Then kick
    /// off a fresh network fetch. On success, swap in the new data
    /// and rewrite the cache. On failure, keep whatever the cache
    /// gave us (and only surface an error if we have nothing at all).
    private func load() async {
        guard activities.isEmpty else { return }

        if let cached = ActivityCache.load() {
            activities = cached
            initialLoading = false
        }

        do {
            let fresh = try await APIClient.shared.listActivities()
            activities = fresh
            errorMessage = nil
            ActivityCache.save(fresh)
        } catch {
            if activities.isEmpty {
                errorMessage = error.localizedDescription
            }
            // else: keep the cached list visible, no error UI
        }
        initialLoading = false
    }

    /// Pull-down refresh. Keep stale data on failure so the user
    /// never sees a spontaneously-empty list mid-session. Writes
    /// through to the cache on success so the next cold start
    /// reflects the most recent fetch.
    private func refresh() async {
        do {
            let fresh = try await APIClient.shared.listActivities()
            activities = fresh
            errorMessage = nil
            ActivityCache.save(fresh)
        } catch {
            // swallow — refreshable has no error UI
        }
    }
}

struct WatchErrorView: View {
    let message: String

    var body: some View {
        VStack(spacing: 6) {
            Image(systemName: "exclamationmark.triangle")
                .font(.system(size: 22))
                .foregroundStyle(.secondary)
            Text("Could not load")
                .font(.headline)
            Text(message)
                .font(.caption2)
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)
                .lineLimit(3)
        }
        .padding(.horizontal, 8)
    }
}

File: android/wear/src/main/java/com/vivianweidai/science/wear/OlympiadsRootView.kt
package com.vivianweidai.science.wear

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.wear.compose.material.CircularProgressIndicator
import androidx.wear.compose.material.MaterialTheme
import androidx.wear.compose.material.Text
import androidx.wear.compose.navigation.SwipeDismissableNavHost
import androidx.wear.compose.navigation.composable
import androidx.wear.compose.navigation.rememberSwipeDismissableNavController
import com.vivianweidai.science.core.api.ApiClient
import com.vivianweidai.science.core.grouping.ActivityGrouping
import com.vivianweidai.science.core.model.Activity

/**
 * Top-level wear screen. Owns the activity list, the current subject
 * filter, and the loading/error state.
 *
 * Differs from the phone OlympiadsView in two ways:
 * - Subject filter is persisted via SharedPreferences so it sticks
 *   across launches. Watch gets opened many times a day; randomizing
 *   would be actively annoying (mirrors iOS `@AppStorage` choice).
 * - On cold start, we synchronously hydrate from [ActivityCache]
 *   before the network call returns, so the first frame is useful
 *   instead of a spinner. Cached list is overwritten only on fetch
 *   success.
 */
@Composable
fun OlympiadsRootView() {
    val context = LocalContext.current
    val nav = rememberSwipeDismissableNavController()

    var activities by remember { mutableStateOf<List<Activity>>(emptyList()) }
    var storedFilter by remember { mutableStateOf(FilterPrefs.read(context) ?: "") }
    var initialLoading by remember { mutableStateOf(true) }
    var errorMessage by remember { mutableStateOf<String?>(null) }

    val effectiveFilter: String? = storedFilter.ifEmpty { null }

    LaunchedEffect(Unit) {
        if (activities.isEmpty()) {
            ActivityCache.load(context)?.let {
                activities = it
                initialLoading = false
            }
            runCatching { ApiClient.shared.listActivities() }
                .onSuccess { fresh ->
                    activities = fresh
                    errorMessage = null
                    ActivityCache.save(context, fresh)
                }
                .onFailure { e ->
                    if (activities.isEmpty()) {
                        errorMessage = e.message ?: e::class.simpleName
                    }
                    // else: keep the cached list visible, no error UI
                }
            initialLoading = false
        }
    }

    val groups = remember(activities, effectiveFilter) {
        val filtered = ActivityGrouping.filtered(activities, effectiveFilter)
        ActivityGrouping.groupedByYear(filtered)
    }

    SwipeDismissableNavHost(navController = nav, startDestination = "list") {
        composable("list") {
            Box(Modifier.fillMaxSize()) {
                when {
                    initialLoading -> Box(
                        Modifier.fillMaxSize(),
                        contentAlignment = Alignment.Center,
                    ) { CircularProgressIndicator() }

                    activities.isEmpty() && errorMessage != null ->
                        WatchErrorView(errorMessage!!)

                    else -> OlympiadsListView(
                        yearGroups = groups,
                        activeFilter = effectiveFilter,
                        onEntry = { activity ->
                            nav.navigate("detail/${activity.id}")
                        },
                        onFilter = { nav.navigate("filter") },
                    )
                }
            }
        }

        composable("detail/{id}") { back ->
            val id = back.arguments?.getString("id")?.toIntOrNull()
            val a = activities.firstOrNull { it.id == id }
            if (a != null) ActivityDetailView(a)
            else Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                Text("Not found")
            }
        }

        composable("filter") {
            SubjectFilterSheet(
                current = effectiveFilter,
                onSelect = { picked ->
                    storedFilter = picked ?: ""
                    FilterPrefs.write(context, picked)
                    nav.popBackStack()
                },
            )
        }
    }
}

@Composable
fun WatchErrorView(message: String) {
    Box(
        modifier = Modifier.fillMaxSize().padding(horizontal = 12.dp, vertical = 8.dp),
        contentAlignment = Alignment.Center,
    ) {
        androidx.compose.foundation.layout.Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = androidx.compose.foundation.layout.Arrangement.spacedBy(4.dp),
        ) {
            Text("Could not load", style = MaterialTheme.typography.title3)
            Text(
                text = message,
                style = MaterialTheme.typography.caption2,
                color = MaterialTheme.colors.onSurfaceVariant,
                maxLines = 3,
            )
        }
    }
}
